feat(benchmark): Add memory regression tests

WayneFerrao opened the pull request:
[AB#32591](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/32591)

## Description

This WIP PR introduces memory regression testing to ensure that memory usage remains stable across test runs. This is in line with goals to strengthen the overall reliability of the the DDSes by ensuring memory-related issues are proactively caught in testing. 
The new logic would detect regressions and allows for setting baselines dynamically. Regression checking is handled inside benchmarkMemory. Individual test objects pass in `baselineMemoryUsage` and `allowedDeviationBytes`do not need to manually check for regressions. 


#### Key Changes
  - Memory Regression Detection:
  - Compares current memory usage against a predefined baseline.
  - Throws an error if memory usage exceeds the allowed threshold.

WayneFerrao committed:
WIP: Add memory regression tests

WayneFerrao committed:
Merge branch 'main' into memoryRegression

WayneFerrao committed:
updated IMemoryTestObject

WayneFerrao committed:
Added env condition

WayneFerrao committed:
revert package.json

WayneFerrao committed:
Comments

WayneFerrao committed:
Lint fix

chentong7 commented on packages/dds/map/src/test/memory/map.spec.ts:
```diff
@@ -11,6 +14,24 @@ import {
+	try {
+		return JSON.parse(fs.readFileSync(baselineFilePath, "utf8")) as Record<string, number>;
+	} catch {
+		return {};
```
Add a console error? console.error("Error loading baselines:", error);

Josmithr replied:
Based on the function docs, it is okay for the baseline to be missing. Might be worth returning `undefined` in that case though, rather than an empty record.

Josmithr replied:
Also, this will fail if the file is missing, but it can also fail for other reasons (malformed file contents, for example). It might be better to first check if the file exists and return early if it doesn't. If it does exist, we probably don't want to eat errors that occur while reading the file.

chentong7 commented on packages/dds/map/src/test/memory/map.spec.ts:
```diff
@@ -11,6 +14,24 @@ import {
+	const baselines = loadBaselines();
+	baselines[testTitle] = memoryUsage;
+	// eslint-disable-next-line unicorn/no-null
+	fs.writeFileSync(baselineFilePath, JSON.stringify(baselines, null, 2));
```
Specified the encoding as "utf8" in the fs.writeFileSync() method to ensure consistent file writing. Like: fs.writeFileSync(baselineFilePath, JSON.stringify(baselines, null, 2), "utf8");

Josmithr commented on packages/dds/map/src/test/memory/map.spec.ts:
```diff
@@ -43,12 +64,18 @@ describe("SharedMap memory usage", () => {
 			public readonly title = "Create empty map";
 			public readonly minSampleCount = 500;
-
+			public baselineMemoryUsage = loadBaselines()[this.title] ?? 0;
```
readonly?

Josmithr commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -116,6 +116,11 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+	/**
+	 * The baseline memory usage to compare against for the test, which is used to determine if the test regressed.
+	 */
+	baselineMemoryUsage?: number;
```
readonly?

WayneFerrao replied:
Cool, I didn't know you could specify that in the interface

Josmithr commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -116,6 +116,11 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
 	samplePercentageToUse?: number;
+
+	/**
+	 * The baseline memory usage to compare against for the test, which is used to determine if the test regressed.
```
Can we document what it means for this value to not be specified?

Josmithr commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -195,11 +200,14 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
 		category: testObject.category ?? "",
 	};
 
+	const ALLOWED_DEVIATION = 5;
```
Docs please 🙂

Josmithr commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +304,16 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+
+					if (avgHeapUsed > allowedMemoryUsage) {
+						throw new Error(
+							`Memory Regression detected for ${testObject.title}: Used ${avgHeapUsed} bytes, exceeding the baseline of ${allowedMemoryUsage} bytes.`,
```
Nit: `allowedMemoryUsage` isn't the actual baseline, it's the baseline + the allowed deviation. It might be better to express this like
```
							`Memory Regression detected for ${testObject.title}: Used ${avgHeapUsed} bytes, exceeding the baseline of ${args.baselineMemoryUsage} bytes., with an allowed tolerance of {tolerance}.`,
```

WayneFerrao replied:
Good catch, will fix!

Josmithr commented on packages/dds/map/src/test/memory/memory_baseline.json:
```diff
@@ -0,0 +1 @@
+{}
```
Will there be 1 of these files per test? Or a single file with all of the test benchmarks?

Josmithr reviewed (commented):
A couple of high-level points of feedback:
1. We should definitely have at least some usages of this in place before merging
2. We should document (in the benchmark package README) how to use the new kinds of benchmarks in tests.

alexvy86 reviewed (commented):
I don't feel too comfortable with the current approach. Did we consider other alternatives for how to keep track of the baselines? That each test file needs to have (or import) functions to read/write baseline files feels weird to me, and I wonder how things would look if each test could just pass its baseline as an optional parameter to `benchmarkMemory()`, and the tool would take care of the necessary comparisons. Having baselines be part of source control also seems convenient.

Along the same lines, maybe having the allowed deviation be a parameter that each test can define (with a reasonable default) would give us more flexibility? Some tests might be able to live with tighter targets, but others might need wider ones. Memory measurements are notoriously finnicky, so I'd be concerned about having a global variability threshold that could make some tests flaky with little recourse other than making changes in benchmark tool again.

To @Josmithr 's point:

> We should definitely have at least some usages of this in place before merging

I agree, but it'll have to be a 2-step process anyway because changes to benchmark tool need to be published and consumed in the client release group separately. Testing those changes before merging them by locally linking benchmark-tool is a good idea though. One other advantage of having all this be parameters on `benchmarkMemory()` is that we might be able to write unit tests for it in benchmark-tool itself.

WayneFerrao committed:
Abstracted load and save baseline logic to benchmarkMemory and added deviation as optional param

WayneFerrao commented:
@alexvy86 @Josmithr I like the idea of having a central file with all the tests and their baselines. Also moving the saving/loading baselines logic to benchmarkMemory makes sense to avoid duplication. I also added a env variable for gating when we want to save/overwrite the baseline. Thoughts?

alexvy86 commented:
> @alexvy86 @Josmithr I like the idea of having a central file with all the tests and their baselines. Also moving the saving/loading baselines logic to benchmarkMemory makes sense to avoid duplication. I also added a env variable for gating when we want to save/overwrite the baseline. Thoughts?

Still not super convinced about the idea of the file, to be honest 😅 . One immediate problem is that if we key the entries in the file just by test title, we could run into conflicts if two tests in different suites have the same title; maybe `testObject.title` is the "fully qualified name" with all the suites and everything, which would take care of that problem, but then my next question is what will happen if a test changes its name (because of a typo, or its purpose changes). The entry in the file with the old name won't get cleaned up so it'll live in the file forever, I think?

In general, I don't like the idea of needing side-effects (disk writes) for tests to work if not strictly necessary. Having it be partially driven by a an env variable to me seems like it introduces more complexity and things one needs to know (and are not super discoverable) in order to work on this kind of tests. When trying to update the baseline for a given test, one needs to be careful to only run _that_ test, because otherwise we would be potentially updating baselines for all memory tests; it would probably get caught during PR review, but still doesn't seem ideal to me.

I think I'd like to see an argument for why the file is necessary or better than other options. To me the "locality" of being able to look at a test in the source file and right there see (and be able to adjust) its expected baseline and allowed variance, feel super useful in comparison.

WayneFerrao committed:
Merge branch 'main' into memoryRegression

WayneFerrao committed:
Embed baseline & allowed deviation in test

WayneFerrao commented:
@alexvy86 That's a good point. We'd need full suite namespacing to avoid test name collisions ( which still doesn't help with stale data/test name changing). I also agree that having local control would be way better for discoverability and manageability. I've updated the code to reflect this. I've changed the env variable to enable the memory tests. Perhaps a requirement to onboard a memory regression test would be run it x times and commit the average to the `baselineMemoryUsage` variable in the test for that DDS?

WayneFerrao committed:
Remove unused imports

WayneFerrao committed:
Fixed linting issues

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -116,6 +118,17 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+
+	/**
+	 * The baseline memory usage to compare against for the test, which is used to determine if the test regressed.
+	 * If not specified, the test will not be compared against a baseline and will only be run to measure the memory usage.
```
```suggestion
	 * If not specified, the test will not be compared against a baseline and will only be run to measure the memory usage.
	 * @remarks Should be specified in bytes.
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -116,6 +118,17 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+	readonly baselineMemoryUsage?: number;
+
+	/**
+	 * The allowed deviation from the baseline memory usage, as a percentage.
```
```suggestion
	 * The allowed deviation from `baselineMemoryUsage`, as a percentage (e.g. 10 for 10%).
	 * Has no effect if `baselineMemoryUsage` is not specified.
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +310,17 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+						args.baselineMemoryUsage * (1 + args.allowedDeviation / 100);
+					const tolerance = args.baselineMemoryUsage * (args.allowedDeviation / 100);
+
+					if (ENABLE_MEM_REGRESSION && avgHeapUsed > allowedMemoryUsage) {
```
I think it'd be nice if we checked separately for variation over or under. If it goes over, it's probably a regression. If it goes under, it might be a performance improvement that should be accompanied by an update to the baseline? And the error message should make that clear.

alexvy86 replied:
If the calculations are only necessary when `ENABLE_MEM_REGRESSION` is set, it's a bit better to move them inside an if block so they only happen when we're going to use them.

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +310,17 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+
+					if (ENABLE_MEM_REGRESSION && avgHeapUsed > allowedMemoryUsage) {
+						throw new Error(
+							`Memory Regression detected for ${testObject.title}: Used ${avgHeapUsed} bytes, exceeding the baseline of ${args.baselineMemoryUsage} bytes., with an allowed tolerance of ${tolerance} bytes.`,
```
Nit: I think leaving the tolerance as a percent for the error is fine. The exact number of bytes won't matter much.

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -24,6 +24,8 @@ import { timer } from "../timer";
 
 import { supportParentProcess } from "./runner";
 
+const ENABLE_MEM_REGRESSION = process.env.ENABLE_MEM_REGRESSION === "true";
```
If the env variable is necessary for the regression-testing behavior, let's document that in the relevant properties, otherwise people could get confused as to why their tests are not regressing.

I'm also on the fence about the need for the env variable... granted, it gives us flexibility to run them and not fail things, but then when/how do we plan to run tests with this variable set?

alexvy86 commented on packages/dds/map/src/test/memory/map.spec.ts:
```diff
@@ -43,12 +43,18 @@ describe("SharedMap memory usage", () => {
 			public readonly title = "Create empty map";
 			public readonly minSampleCount = 500;
-
+			public readonly baselineMemoryUsage = 0; // Example: Baseline memory usage in bytes
```
Tests with a baseline of 0 are tricky... in practice, the measurements will almost never be a clean 0, they will oscillate around it, and if the baseline is 0, the calculation for allowed deviation is 0, and thus those tests will (in practice) always fail complaining about a regression. I'm not sure if there's a good way to work around that. I think it'd be preferable to throw an error indicating baselines very close to or exactly 0 are unsupported or something along those lines.

Or actually: if we get `allowedDeviation` specified in bytes instead of a percentage directly by the test author, then we can just check that the measurement is within `baseline - allowedDeviation < measurement  < baseline + allowedDeviation`. Yeah, I think I like that the best.

alexvy86 commented on packages/dds/map/src/test/memory/map.spec.ts:
```diff
@@ -43,12 +43,18 @@ describe("SharedMap memory usage", () => {
 			public async run(): Promise<void> {
 				this.map = createLocalMap("testMap");
 			}
+			public afterIteration(): void {
+				if (process.env.SAVE_MEMORY_BASELINE) {
+					const memoryUsage = process.memoryUsage().heapUsed;
+				}
+			}
```
Remove?

alexvy86 reviewed (commented):
I definitely prefer this approach. Left several comments though, I think this still needs a bit more work.

WayneFerrao committed:
Updated variables to use byte measurements. Modified env var logic  and updated to include upper/lower bounds

WayneFerrao commented:
@alexvy86 Those were great suggestions. I decided to keep the env variable to allow devs to locally toggle when testing. In prod, this would be set to true so we don't silence regression errors in CI. I also set the `baselineMemoryUsage` and `allowedDeviationBytes` to -1 if they're not being set in the test, since all the props MemoryTestObjectProps are required but not all the DDS have regression tests set up. 

TODO: follow up task to update `baselineMemoryUsage` and `allowedDeviationBytes` to be required props

WayneFerrao committed:
Add build artifacts; api docs

WayneFerrao committed:
Update api-reports

WayneFerrao committed:
Merge branch 'main' into memoryRegression

WayneFerrao committed:
Update api-reports

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -189,19 +204,26 @@ export interface MemorySampleData {
  * @public
  */
 export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+	// Setting to -1 to indicate that baselineMemoryUsage or allowedDeviationBytes variables are not set.
+	const baselineMemoryUsage = testObject.baselineMemoryUsage ?? -1;
+	const allowedDeviationBytes = testObject.allowedDeviationBytes ?? -1;
```
This also feels like the right place to validate that if either `baselineMemoryUsage` or `allowedDeviationBytes` were provided, the other one must have been provided as well. It doesn't make sense to have either without the other, right? Erroring out in that scenario seems correct. If you agree, let's also update their docs to make it clear that users should provide both or none.

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +315,32 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+						const avgHeapUsed = heapUsedStats.arithmeticMean;
+						const lowerBound = baselineMemoryUsage - args.allowedDeviationBytes;
+						const upperBound = baselineMemoryUsage + args.allowedDeviationBytes;
+						const tolerance = (args.allowedDeviationBytes / baselineMemoryUsage) * 100;
```
This can result in a division by 0. I know I said before that keeping the tolerance as percentage seemed fine, to avoid the conversion; now that we receive it in bytes, I think similar reasoning applies to now keep it in bytes, i.e. just use `allowedDeviationBytes` 😄 .

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +315,32 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+						// Throw errors on regressions/improvements if `ENABLE_MEM_REGRESSION` is set and a warning otherwise.
+						// This allows us to run the tests in CI without failing the build, but still get a warning.
+						if (avgHeapUsed > upperBound) {
+							const message = `Memory Regression detected for ${testObject.title}: Used ${avgHeapUsed} bytes, exceeding the ${tolerance}% tolerance of the baseline of ${args.baselineMemoryUsage} bytes.`;
```
Related to previous comment. And adding quotes around all interpolated values so the message is easily readable if one of the values ends up being blank.
```suggestion
							const message = `Memory Regression detected for test '${testObject.title}'. Used '${avgHeapUsed}' bytes, with baseline '${args.baselineMemoryUsage}' and tolerance of '${allowedDeviationBytes}' bytes.`;
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +315,32 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+							}
+						}
+						if (avgHeapUsed < lowerBound) {
+							const message = `Possible memory improvement detected for ${testObject.title}: Used ${avgHeapUsed} bytes, below the ${tolerance}% allowed tolerance of the baseline of ${args.baselineMemoryUsage} bytes. Consider updating the baseline.`;
```
```suggestion
							const message = `Possible memory improvement detected for test '${testObject.title}'. Used '${avgHeapUsed}' bytes, with baseline '${args.baselineMemoryUsage}' and tolerance of '${allowedDeviationBytes}' bytes. Consider updating the baseline.`;
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -116,6 +118,19 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+	 * If not specified, the test will not be compared against a baseline and will only be run to measure the memory usage.
+	 * @remarks Should be specified in bytes.
+	 */
+	readonly baselineMemoryUsage?: number | undefined;
```
Should be handled by the `?`.
```suggestion
	readonly baselineMemoryUsage?: number;
```

alexvy86 commented on tools/benchmark/api-report/benchmark:
Any ideas why this new file got generated? It shouldn't be here.

alexvy86 reviewed (commented):
Looking better :) . Next batch of comments. It's hard to unit-test changes to the benchmark tool, but I'd like to see evidence of using the changes in a test in the client release group (the one for `map` that I ask below be split to a separate PR is a good candidate for that). How does the output look like when it's within threshold, when it's above, when it's below (just tweak the baseline and deviation to force it to be above/below), with and without the ENV variable.

WayneFerrao committed:
Update to use bytes, added validation for memory values

WayneFerrao committed:
remove file

WayneFerrao committed:
Merge branch 'main' into memoryRegression

WayneFerrao committed:
Updated logging to ensure warnings and errors propagate.

WayneFerrao commented:
Here are some screenshots with example usage in the Sharedmap memory tests.

ENABLE_MEM_REGRESSION set to true with high values passed in
<img width="1194" alt="Screenshot 2025-04-29 at 6 55 02 PM" src="https://github.com/user-attachments/assets/50988c0b-6154-43b8-8eec-b899b04ff766" />

ENABLE_MEM_REGRESSION set to true with low values passed in
<img width="893" alt="Screenshot 2025-04-29 at 6 55 50 PM" src="https://github.com/user-attachments/assets/1d370062-5ead-4185-92d0-ab37de164693" />

ENABLE_MEM_REGRESSION set to false with high values passed in. Prints warning and continues with test.
<img width="1199" alt="Screenshot 2025-04-29 at 7 01 17 PM" src="https://github.com/user-attachments/assets/7d14fa00-d7e5-42ca-ac3d-ee1145e4befe" />

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -108,14 +115,27 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+	/**
+	 * The baseline memory usage to compare against for the test, which is used to determine if the test regressed.
+	 * If not specified, the test will not be compared against a baseline and will only be run to measure the memory usage.
+	 * @remarks Should be specified in bytes.
```
This is what I had in mind for documenting the env variable. People writing memory tests and using the API cannot see the docs for the `const` in this file, but they can see the ones for these properties, so this is where we can best communicate to them how to use `ENABLE_MEM_REGRESSION`.
```suggestion
	 * @remarks
	 * Should be specified in bytes.
	 * If `ENABLE_MEM_REGRESSION=1` in the environment, a test whose memory usage falls outside `baselineMemoryUsage +- allowedDeviationBytes` will be marked as failed.
	 * Otherwise, a warning is printed to the console.
```

alexvy86 replied:
Re-opening this comment. I see the updated docs in the property below but not in this one.

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +344,43 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+								// eslint-disable-next-line unicorn/no-process-exit
+								process.exit(1);
+							} else {
+								process.stdout.write(message);
```
Try this and confirm it does what I expect? I know we use `chalk` in other places in the tool already, it would be nice if these warnings jumped more in the output. Same for the other warning below.
```suggestion
								process.stdout.write(chalk.yellow(message));
```

"!")

WayneFerrao replied:
Yep, chalk works! Warnings show up in yellow

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +344,43 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+								args.baselineMemoryUsage
+							}' and tolerance of '${allowedDeviationBytes} bytes.\n`;
+							if (ENABLE_MEM_REGRESSION) {
+								process.stderr.write(message);
+								// eslint-disable-next-line unicorn/no-process-exit
+								process.exit(1);
```
Thanks for the screenshots! They did showcase a problem we need to fix: when `ENABLE_MEM_REGRESSION=1` we still need to see the table output like when it is set to 0, including any other tests that might have run, and the regressed test should be marked as failed in there. Right now this makes it so as soon as a test is determined to have regressed, we just terminate the whole process, run no further tests, and get no output for anything else.

I think just `throw new Error(message)` here should fix it... all this is already wrapped in a try-catch that sets `benchmarkStats = { error: (error as Error).message };` and should result in the expected behavior (the specific test ending up as failed in the output).

WayneFerrao replied:
@alexvy86 I had tried this. `throw new Error(message)` just causes the test to fail silently since it's the child process throwing the error because of this line. 

```typescript
if (result.stderr !== "") {
	assert.fail(`Child process logged errors: ${result.stderr}`);
}
```

From what I can tell, Mocha catches the error but doesn't surface it back to the parent process.

alexvy86 replied:
There's something we need to look into in that case though, because I'm almost certain we follow a similar pattern for the other kinds of benchmarks and that's how they behave. The child process should be printing errored tests to `stdout` somehow, for the parent process to then aggregate results and print the output.

alexvy86 commented on tools/benchmark/src/mocha/runner.ts:
```diff
@@ -110,11 +110,16 @@ export function supportParentProcess<
 			}
 
 			if (result.stderr !== "") {
+				console.error(result.stderr);
```
Do we need this? The line right below seems like it would print it already.

alexvy86 commented on tools/benchmark/src/mocha/runner.ts:
```diff
@@ -110,11 +110,16 @@ export function supportParentProcess<
 			const childProcess = await import("node:child_process");
 			const result = childProcess.spawnSync(command, childArgs, { encoding: "utf8" });
 
+			if (result.stdout) {
+				console.log(result.stdout);
+			}
```
In light of the changes above, we can't do this. Further down we depend on `stdout` from child processes having nothing but a JSON payload; that's basically the interface/contract between parent and child processes. Adding plaintext output to that will probably break something else. Try to see if there's a way of communicating the warning within the existing payload that child processes write to their output. If not, let's discuss how/if we can keep the warning.

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +344,43 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+								testObject.title
+							}': Used '${avgHeapUsed.toPrecision(6)}' bytes, with baseline'${
+								args.baselineMemoryUsage
+							}' and tolerance of '${allowedDeviationBytes} bytes.\n`;
```
Oops, I missed this in a previous suggestion
```suggestion
							}' and tolerance of '${allowedDeviationBytes}' bytes.\n`;
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -296,6 +344,43 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+								testObject.title
+							}'. Used '${avgHeapUsed.toPrecision(6)}' bytes with baseline '${
+								args.baselineMemoryUsage
+							} and tolerance of '${allowedDeviationBytes} bytes. Consider updating the baseline.\n`;
```
```suggestion
							}' and tolerance of '${allowedDeviationBytes}' bytes. Consider updating the baseline.\n`;
```

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -108,14 +115,27 @@ export interface MemoryTestObjectProps extends MochaExclusiveOptions, Titled, Be
+	readonly baselineMemoryUsage?: number;
+
+	/**
+	 * The allowed deviation from the `baselineMemoryUsage`, measured in bytes.
+	 * Has no effect if `baselineMemoryUsage` is not specified.
```
```suggestion
	 * The allowed deviation from the `baselineMemoryUsage`, measured in bytes.
	 * @remarks
	 * Has no effect if `baselineMemoryUsage` is not specified.
	 * If `ENABLE_MEM_REGRESSION=1` in the environment, a test whose memory usage falls outside `baselineMemoryUsage +- allowedDeviationBytes` will be marked as failed.
	 * Otherwise, a warning is printed to the console.
```

WayneFerrao committed:
Updated logging with chalk, fixed logs, moved logic.

WayneFerrao committed:
Restore logic

alexvy86 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -320,6 +371,41 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
+								process.stderr.write(message);
+								throw new Error(message);
+							} else {
+								process.stdout.write(chalk.yellow(message));
```
This one still concerns me. Child processes writing anything to stdout (other than the final test results) feels like it would break the communication between parent and child. Let's double check what happens when things are working (i.e. using `node_modules/` in the reporter path when pnpm-linked) and we hit this line by having a test that goes above/below threshold. I suspect we'll need to remove it, and probably also remove the write to stderr, since the `throw` should already be communicating the problem all the way to the output table.

alexvy86 replied:
Update: we realized that extra lines in a child process' output don't break things, as long as the child process also outputs one line with the JSON payload that it communicates back to the parent process, so this is not a problem.

alexvy86 commented on tools/benchmark/package.json:
```diff
@@ -41,7 +41,8 @@
 		"mocha-multi-reporters": "^1.5.1",
-		"moment": "^2.21.0"
+		"moment": "^2.21.0",
+		"picocolors": "^1.1.1"
```
If we end up removing the stdout/stderr writes, I think this becomes unnecessary.

WayneFerrao committed:
Remove unneeded package and added comments

WayneFerrao committed:
Add comment

WayneFerrao committed:
Updated comment for baselineMemoryUsage

chentong7 commented on tools/benchmark/src/mocha/memoryTestRunner.ts:
```diff
@@ -320,6 +373,41 @@ export function benchmarkMemory(testObject: IMemoryTestObject): Test {
 						formattedValue: prettyNumber(runs, 0),
 					};
+
+					if (baselineMemoryUsage >= 0 && allowedDeviationBytes >= 0) {
```
Looks like there are some duplications. More readable way is like:
```suggestion
					if (avgHeapUsed > upperBound) {
        reportMemoryIssue(
            `Memory regression detected for test '${testTitle}': Used '${avgHeapUsed.toPrecision(
                6,
            )}' bytes, baseline '${baselineMemoryUsage}', tolerance '${allowedDeviationBytes}' bytes.\n`,
        );
    } else if (avgHeapUsed < lowerBound) {
        reportMemoryIssue(
            `Possible memory improvement detected for test '${testTitle}': Used '${avgHeapUsed.toPrecision(
                6,
            )}' bytes, baseline '${baselineMemoryUsage}', tolerance '${allowedDeviationBytes}' bytes. Consider updating the baseline.\n`,
        );
    }
```

chentong7 replied:
```suggestion
function reportMemoryIssue(message: string): void {
    if (ENABLE_MEM_REGRESSION) {
        throw new Error(message);
    } else {
        process.stdout.write(chalk.yellow(message));
    }
}
```

WayneFerrao replied:
Great idea, thanks!

WayneFerrao committed:
update CHANGELOG

WayneFerrao committed:
Refactor memory issue reporting to separate function

WayneFerrao committed:
Merge branch 'main' into memoryRegression

alexvy86 reviewed (approved):
🚀 . Once it merges, let's make sure to do a release of this package and consume it in the client release group so we can start leveraging the feature :)
